File: day2.jl
struct Password

    cmin :: Int
    cmax :: Int
    char :: Char
    pass :: String

end

function read_passwords(filename)

    passwords = Password[]
    open(filename) do f
         for line in eachline(f)
             data = split(line)
             cmin, cmax  = parse.(Int, split(data[1],"-"))
             char = first(data[2])
             pass = data[3]
             push!(passwords, Password(cmin, cmax, char, pass))
         end
         return passwords
    end

end

function check1(p) 
    p.cmin <= count(c->(c==p.char), p.pass) <= p.cmax
end


function check2(p) 
    ans = ((p.pass[p.cmin] == p.char && p.pass[p.cmax] != p.char) ||
          (p.pass[p.cmin] != p.char && p.pass[p.cmax] == p.char) )
    return ans
end

passwords = read_passwords("input2.txt")
println(sum(check1.(passwords)))
println(sum(check2.(passwords)))

File: day2.py
from collections import namedtuple

Password = namedtuple("Password", "cmin cmax char, pawd")

def read_passwords(filename):

    passwords = []
    with open(filename) as f:
         for line in f:
             d1, d2, d3 = line.split()
             cmin, cmax  = list(map(int, d1.split("-")))
             char = d2[0]
             pawd = d3
             passwords.append(Password(cmin, cmax, char, pawd))
         return passwords

def check1(p):
     return p.cmin <= p.pawd.count(p.char) <= p.cmax

 
def check2(p):
     return (p.pawd[p.cmin-1] == p.char) ^ (p.pawd[p.cmax-1] == p.char)

passwords = read_passwords("input2.txt")
print(sum(map(check1, passwords)))
print(sum(map(check2, passwords)))
